File: uva/stl/400.py
import io
import sys
import collections


# Simulate the redirect stdin.
if len(sys.argv) > 1:
    filename = sys.argv[1]
    inp = ''.join(open(filename, "r").readlines())
    sys.stdin = io.StringIO(inp)

def myPrint(s, M , extra):
    print(s, end='')
    for _ in range(M-len(s)):
        print(extra, end='')


def main():
    try:
        while(1):
            n = int(input())
            filenames=[]
            M = 0
            for i in range(n):
                filenames.append(input())
                M= max(M,len(filenames[i]))

            cols = (60 - M)//(M+2) +1
            rows = (n-1) // cols +1

            filenames.sort()
            print("------------------------------------------------------------")
            for r in range(rows):
                for c in range(cols):
                    i = c* rows +r
                    if i < n :
                        myPrint(filenames[i], M if c == cols-1 else M+2, ' ' )
                print()
                


    except EOFError:
        return 0
        


if __name__ == "__main__":
    main()

File: uva/stl/400.cpp
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <cmath>
using namespace std;

// 输出字符串s，长度不足len时补字符extra
void print(const string& s, int len, char extra) {
  cout << s;
  for(int i = 0; i < len-s.length(); i++)
    cout << extra;
}

int main(){
    int n;
    while(scanf("%d",&n) == 1) {
        vector<string> filenames;
        filenames.resize(n);
        int M = 0;
        for (int i =0; i <n; i++) {
            cin >> filenames[i];
            M = max(M, (int)filenames[i].length());
        }
        sort(filenames.begin(),filenames.end());
        printf("------------------------------------------------------------\n");

        int cols = (60 - M)/(M+2) +1;
        int rows = (n-1) / cols +1;

        for (int r =0; r < rows; r++) {
            for (int c =0; c< cols; c++) {
                int idx = c * rows +r;
                if (idx < n) {
                    print(filenames[idx],  c == cols-1 ? M : M+2 , ' ');
                }
            }
            cout << "\n";
        }
 
    }
}